Introduce System Status and Use For Boot Detection

ejoerns opened the pull request:
This introduces support for writing a global rauc state file in key-value format, located in the data-directory.
For now, the file will be used to store the system's boot ID:

    [system]
    boot-id=cfb75bf4-e678-42e7-8dc9-14fed56baf1a

With the boot-id information from the state file, the service can determine if the system was (re)booted.

This works by comparing the current boot ID with the ID stored in the status file.
If they differ, we know a reboot occurred and then store the current boot ID in the status file.
A subsequent restart of the service would the see the current and the stored boot ID being equal and just report a service restart.

We emit a normal log message for now but the plan is to use this later for dedicated logging of reboots.

jluebbe commented on src/main.c:
```diff
@@ -1390,14 +1390,12 @@ static gchar* r_status_formatter_readable(RaucStatusPrint *status)
-	bootedfrom = r_slot_find_by_device(status->slots, status->bootslot);
-	if (!bootedfrom)
-		bootedfrom = r_slot_find_by_bootname(status->slots, status->bootslot);
+	bootedfrom = r_slot_get_booted(status->slots);
```
If `r_slot_get_booted` returned NULL, that could also be mapped to `"unknown"`.

jluebbe commented on src/context.c:
```diff
@@ -315,6 +316,15 @@ static gboolean r_context_configure_target(GError **error)
+	/* load global state */
+	g_clear_pointer(&context->statepath, g_free);
+	g_clear_pointer(&context->global_state, r_global_state_free);
+	context->statepath = g_build_filename(context->config->data_directory, "state", NULL);
```
`data_directory` can be unset. Perhaps move it inside the `if` above?

Also, `state` as a filename seems too generic. Perhaps something mentioning persistence?

ejoerns replied:
We discussed that due to atomicity using the existing status file should be preferred.
Then we still need to check for per-slot not being enabled.

I'll implement it this way but rethinking your proposal I am not sure if atomicity plays any role since the update of the slot status and the update of the system status would never happen at the same time I guess.

jluebbe commented on src/main.c:
```diff
@@ -2011,6 +2011,23 @@ static gboolean service_start(int argc, char **argv)
+		g_free(r_context()->global_state->boot_id);
+		r_context()->global_state->boot_id = g_strdup(r_context()->boot_id);
+
+		if (!r_global_state_save(r_context()->statepath, r_context()->global_state, &ierror)) {
+			g_warning("Failed saving global state");
+			g_clear_error(&ierror);
+		}
```
I find passing data from `r_context()` into a different module strange. Perhaps this should a helper in `context.h` or `r_global_state_save` should access the context directly?

ejoerns replied:
We discussed that (unlike for the load function which gets called before having context set up) passing path and state object explicitly is not required here. I will check if we can also use a wrapper method in `context.c` to hide the direct call to slot saving method.

jluebbe replied:
You decided against the wrapper, but with out the explicit parameters, I'm fine with this.

jluebbe reviewed (changes requested):
I'm not convinced that a separate file is the best approach, as it means that we can't have atomic updates to the global state and the slot status.

Also, we should probably be explicit about the requirement for the data directory.

ejoerns commented on src/global_state.c:
```diff
@@ -145,6 +146,17 @@ gboolean determine_slot_states(GError **error)
 			}
 
+			/* Create and add dummy external slot */
+			ext_dummy = g_new0(RaucSlot, 1);
+			ext_dummy->name = g_strdup("external");
+			ext_dummy->sclass = g_strdup("external");
+			ext_dummy->type = g_strdup("virtual");
```
I still like the idea but I will omit it for the next PR update first of all

jluebbe commented on src/config_file.c:
```diff
@@ -631,7 +631,9 @@ gboolean load_config(const gchar *filename, RaucConfig **config, GError **error)
 		} else {
-			g_message("Config option 'statusfile=<path>/per-slot' unset, falling back to per-slot status");
+			g_message("No data directory or status file set, falling back to per-slot status.\n"
+					"This will limit supported features.\n"
+					"Consider setting 'data-directory=<path>' or 'statusfile=<path>/per-slot' explicitly.");
```
`per-slot` will still limit supported features. Perhaps the limited features message should be triggered by `per-slot` only?

ejoerns replied:
So we should probably just omit the limited feature line. For `per-slot` case in general I already I introduced the follow-up message that says

> Using per-slot statusfile. System status information not supported!

I'll also adapt the commit message accordingly

jluebbe commented on src/context.c:
```diff
@@ -315,6 +316,15 @@ static gboolean r_context_configure_target(GError **error)
+	if (g_strcmp0(context->config->statusfile_path, "per-slot") != 0) {
+		g_clear_pointer(&context->system_status, r_system_status_free);
+		context->system_status = g_new0(RSystemStatus, 1);
+		if (!r_system_status_load(context->config->statusfile_path, context->system_status, &ierror)) {
```
`ierror` needs to be cleared on in the error case, otherwise the next function can't use it.

ejoerns replied:
Will fix

jluebbe commented on src/context.c:
```diff
@@ -315,6 +316,15 @@ static gboolean r_context_configure_target(GError **error)
+		g_clear_pointer(&context->system_status, r_system_status_free);
+		context->system_status = g_new0(RSystemStatus, 1);
+		if (!r_system_status_load(context->config->statusfile_path, context->system_status, &ierror)) {
+			g_message("Failed to load global state: %s\n", ierror->message);
```
```suggestion
			g_message("Failed to load system status: %s\n", ierror->message);
```

jluebbe commented on include/status_file.h:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Save sytem status from central status file.
+ *
+ * Note that filename and state are passed explicitly here since the method is
+ * desinged to be called during context setup where we cannot access context,
+ * yet.
+ *
+ * @param filename File name to load state from
+ * @param state RSystemStatus to update from file
+ * @param[out] error Return location for a GError, or NULL
+ *
+ * @return
+ */
+gboolean r_system_status_load(const gchar *filename, RSystemStatus *state, GError **error)
```
```suggestion
 * Note that filename and status are passed explicitly here since the method is
 * designed to be called during context setup where we cannot access context,
 * yet.
 *
 * @param filename File name to load status from
 * @param status RSystemStatus to update from file
 * @param[out] error Return location for a GError, or NULL
 *
 * @return
 */
gboolean r_system_status_load(const gchar *filename, RSystemStatus *status, GError **error)
```

jluebbe commented on include/status_file.h:
```diff
@@ -0,0 +1,96 @@
+G_GNUC_WARN_UNUSED_RESULT;
+
+/**
+ * Save sytem status to central status file.
```
```suggestion
 * Save system status to central status file.
```

jluebbe commented on include/status_file.h:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Free system status.
+ *
+ * @param state RSystemStatus to free
+ */
+void r_system_status_free(RSystemStatus *state);
```
```suggestion
 * @param status RSystemStatus to free
 */
void r_system_status_free(RSystemStatus *status);
```

jluebbe commented on src/main.c:
```diff
@@ -2013,6 +2014,29 @@ static gboolean service_start(int argc, char **argv)
 	}
 
+	if (r_context()->system_status) {
+		/* Boot ID-based boot vs service restart detection */
```
```suggestion
		/* Boot-ID-based system reboot vs service restart detection */
```

ejoerns commented on src/status_file.c:
```diff
@@ -320,6 +323,28 @@ static gboolean save_slot_status_globally(GError **error)
+			g_clear_error(&ierror);
+		} else {
+			g_propagate_error(error, ierror);
+			return FALSE;
```
Unsure. Having it broken in the save method would mean that it broke after loading it. Do we assume this might happen?

ejoerns commented:
Also moved `config_file_statusfile_missing` check back to `test/config_file.c` since it is more of a config-related test.

ejoerns commented:
Rebased onto master to resolve conflict in `src/mark.c`.

ejoerns commented on src/status_file.c:
```diff
@@ -0,0 +1,471 @@
+{
+	g_autoptr(GKeyFile) key_file = NULL;
+	GError *ierror = NULL;
+	g_auto(GStrv) groups = NULL;
```
Fixed

jluebbe commented on include/status_file.h:
```diff
@@ -0,0 +1,96 @@
+} RSystemStatus;
+
+/**
+ * Save sytem status from central status file.
```
```suggestion
 * Load system status from central status file.
```

jluebbe commented on src/status_file.c:
```diff
@@ -308,9 +308,63 @@ static gboolean save_slot_status_locally(RaucSlot *dest_slot, GError **error)
+ *
+ * @return newly-allocated GKeyFile or NULL on error
+ */
+static GKeyFile* key_file_init_from_shared_status_file(const gchar *path, const gchar *keep_prefix, GError **error)
```
The path argument seems unused. Was this intended to be used in unit tests?

ejoerns replied:
I guess this is a remnant from refactoring. Cannot remember having something in mind for it. Removed.

jluebbe commented on src/context.c:
```diff
@@ -315,6 +316,16 @@ static gboolean r_context_configure_target(GError **error)
+		g_clear_pointer(&context->system_status, r_system_status_free);
+		context->system_status = g_new0(RSystemStatus, 1);
+		if (!r_system_status_load(context->config->statusfile_path, context->system_status, &ierror)) {
+			g_message("Failed to load sytem status: %s", ierror->message);
```
```suggestion
			g_message("Failed to load system status: %s", ierror->message);
```

jluebbe commented on src/status_file.c:
```diff
@@ -0,0 +1,470 @@
+		} else {
+			g_autofree gchar *broken_file = g_strconcat(r_context()->config->statusfile_path, ".broken", NULL);
+			/* Allow recovering from broken status file */
+			g_warning("Failed loading status file: %s\n", ierror->message);
+			g_warning("Moving status file to %s and re-creating it.", broken_file);
```
```suggestion
			g_warning("Failed to load status file: %s\n", ierror->message);
			g_warning("Will move status file to %s and re-create it.", broken_file);
```

jluebbe reviewed (changes requested):
We also need test cases for the newly added functions including the broken file handling.

For the shared handling, perhaps add add save_/load_ of the other kind to see if something is damaged?

ejoerns commented:
I've addressed the review feedback and added a bunch of test cases.

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,366 @@
+	system_status = g_new0(RSystemStatus, 1);
+	g_assert_nonnull(system_status);
+
+	/* assert error-free loading */
```
This comment doesn't match the code.

ejoerns replied:
May contain traces of copy-paste :smirk:

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,366 @@
+	pathname = g_build_filename(fixture->tmpdir, "system_only.conf", NULL);
+	g_assert_nonnull(pathname);
+
+	r_context_conf()->configpath = g_strdup("test/test-global.conf");
+	r_context()->config->statusfile_path = pathname;
+	r_context()->system_status->boot_id = g_strdup("e02a2afe-cf45-4d50-a3f3-c223ca0f480a");
```
Perhaps use `replace_strdup`?

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,366 @@
+	pathname = write_tmp_file(fixture->tmpdir, "existing_config.conf", cfg_file, NULL);
+	g_assert_nonnull(pathname);
+
+	r_context_conf()->configpath = g_strdup("test/test-global.conf");
+	r_context()->config->statusfile_path = pathname;
+	r_context()->system_status->boot_id = g_strdup("e02a2afe-cf45-4d50-a3f3-c223ca0f480a");
```
Perhaps use `replace_strdup`?

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,366 @@
+	r_context()->config->statusfile_path = pathname;
+	r_context()->system_status->boot_id = g_strdup("e02a2afe-cf45-4d50-a3f3-c223ca0f480a");
+
+	g_message("EXPECT FORM NOW!");
```
Leftover from debugging?

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,366 @@
+	g_test_add("/status-file/save-system-status-broken-existing-slot-status", ConfigFileFixture, NULL,
+			status_file_fixture_set_up_global,
+			status_file_test_save_system_status_broken_existing_slot_status,
+			status_file_fixture_tear_down);
```
Do we need the other side (save slot status in a file with existing system status) as well?

ejoerns replied:
Thought about this, too. Since it uses the same helper it should not fail the other way round, but, maybe better safe than sorry. Will add a test for this, too.

jluebbe commented on test/meson.build:
```diff
@@ -16,6 +16,7 @@ tests = [
   'utils',
   'install',
   'service',
+  'status_file',
```
Add this new file to https://github.com/rauc/rauc/blob/master/.github/workflows/sanitizers.yml, otherwise we loose address sanitizer coverage for these tests.

ejoerns replied:
Ack. Forgot that it is a separate list. Will add.

ejoerns commented:
Collected a couple of further fixes for tests while working on the review feedback:

* renamed 'cfg/.conf' to 'status/.raucs' for test status file creation
* Fixed typos
* Renamed `ConfigFileFixture` to `StatusFileFixture`
* Use `configpath` initialization from fixture (was replicated in each test)
* removed unused `pathname` in `status_file_test_save_system_status_per_slot()`
* Added missing cleanup of `system_status` in `r_context_clean()`
* Fixed various memory leaks by using `replace_strdup()`
* Fixed further memory leaks in strings, GError, GKeyFile, etc.

jluebbe commented on test/status_file.c:
```diff
@@ -0,0 +1,406 @@
+	g_assert_true(g_strv_contains((const gchar * const *)groups, "system"));
+}
+
+/* Creates and saves system status into file with existing slot status */
+static void status_file_test_save_slot_status_existing_system_status(StatusFileFixture *fixture,
+		gconstpointer user_data)
```
```suggestion
/* Creates and saves slot status into file with existing system status */
static void status_file_test_save_slot_status_existing_system_status(StatusFileFixture *fixture,
		gconstpointer user_data)
```

jluebbe reviewed (changes requested):
Just a minor issue with the comment. Please rebase and this can go in.

ejoerns commented:
Fixed comment. No rebase, yet.

ejoerns commented:
Rebased onto master now and fixed a faulty-fixuped hunk so that it is also now fully bisectable again.
